### PR TITLE
[RFC] Fix clang analysis warnings. (10)

### PR DESCRIPTION
**Pull request by elmart**
Fix existing 2 warnings.

### CODE REVIEW

**Comment by splinterofchaos** [comment]
Needs some parenthesise, but It all looks good.  `init_history` works in a wierd way, though. Is there any reason we have to maintain the shape of the circular list? I would think a simpler algorithm would be:
```c
ssize_t j = hisidx[type];
ssize_t i = newlen - (newlen < hislen ? 1 : newlen - hislen - 1);
ssize_t end = i;
for (; j != hisidx[type]; --i) {
  if (i >= 0) {
    temp[i] = history[type][j];
  } else {
    xfree(history[type][j];
    history[type][j].hisstr = NULL;
  }
  if (--j < 0) {
    j = hislen - 1;
  }
}
for (; end < newlen; end++) {
  clear_hist_entry(&temp[end]);
}
```

**Comment by elmart** [comment]
I wondered the same thing. I didn't see a reason not to reorder the circular array also in the "growing" case. So we can probably change that into something clearer. As I wasn't sure, though, and didn't have time to check if there was any reason, I just decided to add comments, to make it easier to understand, as it was tricky to get it for me.

Code above, though, wouldn't work. For loop is never entered. You'd need to modify the end of the first loop to:
```
if (--j == hisidx[type]) {
  break;
}
else if (j < 0) {
  j = hislen - 1
}
```
for it to work.

**Comment by elmart** [comment]
And removing condition from the for, I mean.

**Comment by elmart** [comment]
Also lines
```
  ssize_t i = newlen - (newlen < hislen ? 1 : newlen - hislen - 1);
  ssize_t end = i;
```
should be 
```
 ssize_t i = newlen - (newlen < hislen ? 1 : newlen - hislen + 1);
 ssize_t end = i +1;
```
IMO.

**Comment by elmart** [comment]
You're right that, being rigorous, `ssize_t` only guarantees negative value -1.
As, in this case, we are representing values restricted to 0 <= value <= 10000, we can definitely use int.
Now, instead of asserting `p_hi <= 10000`, which is already done by `set_num_option()` at option.c:5355, 
I will do the normal in-bounds assert to be sure `p_hi` value can be casted to int. That way, if `p_hi` limits are changed later, we'll only need to review this if new limits don't fit into an int.

**Comment by elmart** [comment]
Repushed with suggested changes.

File: src/nvim/ex_getln.c [hunk 1]
```diff
@@ -4169,58 +4170,47 @@ static char_u *get_history_arg(expand_T *xp, int idx)
...
-    ssize_t j;
+  assert(p_hi >= 0 && p_hi <= INT_MAX);
+  int newlen = (int)p_hi;
+  assert(newlen != 0 || newlen <= hislen);
 
-    // adjust the tables
+   // If history tables size changed, reallocate them.
+   // Tables are circular arrays (current position marked by hisidx[type]).
+   // On copying them to the new arrays, we take the chance to reorder them.
+  if (newlen != hislen) {
```
Comment (dougblack) [inline src/nvim/ex_getln.c]: These comments aren't *quite* lined up with the `if` statement below.

**Comment by doppioandante** [comment]
@elmart I would say the first one because memcpy makes apparent what's being done. Index initialization is more convoluted in the first code, but old_idx is harder to track in the second code.

File: src/nvim/ex_getln.c (hunk 1 above)
Comment (elmart) [inline src/nvim/ex_getln.c]: :+1:

**Comment by elmart** [comment]
Repushed.

**Comment by elmart** [comment]
I was tempted to remove the block
```c
      // clear remaining space, if any
      for (int i = j < 0 ? 0 : MIN(newlen, oldlen); i < newlen; i++) {
        clear_hist_entry(&temp[i]);
      }
```
by using `xcalloc` instead of `xmalloc`. But that would be inefficient in the general case, so I leave it as is.

**Comment by elmart** [comment]
@doppioandante Yes, I also find it more easily understandable, if meaning of calculation is properly explained. I've added a lot of comments for it to be crystal-clear.

**Comment by elmart** [comment]
Repushed.

**Comment by elmart** [comment]
Repushed. Changed also `for (int type = 0; type < HIST_COUNT; ++type) {` to suffix increment, and removed last semicolon from `copy as much entries as they fit to new table, reordering them;`.

**Comment by splinterofchaos** [comment]
> I was tempted to remove the block [...code...] But that would be inefficient in the general case...

What about
```c
if (oldlen < newlen) memset(temp + oldlen, 0, newlen - oldlen);
```

**Comment by elmart** [comment]
Yes, I like that. :+1: 
It matches the spirit of "acting in batches" of code in this function.

**Comment by elmart** [comment]
It does have to take into account the "empty" case, though (j < 0), where we have to initialize everything.

**Comment by elmart** [comment]
Repushed. Replaced space clearing loop with memset call. As an added bonus, now we don't need the assert anymore, as we're not calling clear_hist_entry() anymore.

File: src/nvim/ex_getln.c [hunk 2]
```diff
@@ -4169,58 +4170,60 @@ static char_u *get_history_arg(expand_T *xp, int idx)
...
+          xfree(history[type][i].hisstr);
+          history[type][i].hisstr = NULL;
         }
-        xfree(history[type]);
-        history[type] = temp;
       }
+
+      // clear remaining space, if any
+      if (newlen > oldlen) {
+        memset(temp + (j < 0 ? 0 : oldlen), 0, newlen - oldlen);
```
Comment (oni-link) [inline src/nvim/ex_getln.c]: `memset()`  expects number of bytes, not number of elements.

Comment (elmart) [inline src/nvim/ex_getln.c]: Oops, sorry for that. Test were failing, in fact. You shortened my bug-finding. Thanks!

**Comment by elmart** [comment]
Fixed problem in memset and repushed.

File: src/nvim/ex_getln.c [hunk 3]
```diff
@@ -4169,58 +4170,61 @@ static char_u *get_history_arg(expand_T *xp, int idx)
...
+        for (int i = i1 + l1; i < i2; i++) {
+          xfree(history[type][i].hisstr);
+          history[type][i].hisstr = NULL;
         }
-        xfree(history[type]);
-        history[type] = temp;
       }
+
+      // clear remaining space, if any
+      if (newlen > oldlen) {
```
Comment (oni-link) [inline src/nvim/ex_getln.c]: Clearing could also be necessary for `newlen < oldlen` (`j == -1`, history never had any entries and we reduce `hislen`).

Comment (elmart) [inline src/nvim/ex_getln.c]: Aaargggh! It's incredible how tricky this gets (seeming a simple thing like it looks like). ;-)

**Comment by elmart** [comment]
Fixed memset problem when j == -1 && newlen < oldlen and repushed.

File: src/nvim/ex_getln.c [hunk 4]
```diff
@@ -4169,58 +4170,59 @@ static char_u *get_history_arg(expand_T *xp, int idx)
...
+          xfree(history[type][i].hisstr);
+          history[type][i].hisstr = NULL;
         }
-        xfree(history[type]);
-        history[type] = temp;
       }
+
+      // clear remaining space, if any
+      int l3 = j < 0 ? 0 : MIN(newlen, oldlen);  // number of copied entries
+      memset(temp + l3, 0, (size_t)MAX(newlen - l3, 0) * sizeof(*temp));
```
Comment (oni-link) [inline src/nvim/ex_getln.c]: Why the `MAX(...)`? `(newlen - l3)` should give the same result.

**Comment by elmart** [comment]
I was just trying to ensure it wasn't negative and didn't realize it couldn't.
Removed spurious MAX and repushed.

**Comment by elmart** [comment]
Ok guys, you ok with this?

**Comment by oni-link** [comment]
LGTM.

**Comment by elmart** [comment]
@splinterofchaos?

**Comment by splinterofchaos** [comment]
LGTM :+1:

**Comment by elmart** [comment]
I'm glad with the result of this PR. I think we did a really good job here: we ended up with much more understandable code, and even more performant. Great!

Thanks very much for your insights, your patience, and your time.

**Comment by elmart** [comment]
Ouch, it seems our new shiny code produced a new warning.
It's an easy false positive. I'll fix it in soon.
BTW, I always check before pushing that analyzer doesn't complain. But here we did have real bad luck. My version of clang (which is older than that run by bot) doesn't signal this issue. I can't use the latest clang on OSX, because neovim doesn't compile with it (it's been so for a long time; the issue was recently fixed in clang, but hasn't been included in any public release yet). It's the first time that such a thing happens to me (my version of analyzer not catching an issue that the newer one does).

**Pull request merged**